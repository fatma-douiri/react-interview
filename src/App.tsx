import { FC, useEffect, useMemo, useState } from 'react';

import CategoryFilter from './components/categoryFilter/CategoryFilter';
import Loader from './components/loader/Loader';
import Pagination from './components/pagination/Pagination';
import MoviesList from './features/movies/MoviesList';
import { useMovies } from './hooks/useMovies';
import { Movie, MovieStatusEnum } from './types/movie';

const App: FC = () => {
  const { movies, loadMovies, status, error, currentPage, itemsPerPage, changePage } = useMovies();

  const [filteredMovies, setFilteredMovies] = useState<Movie[]>(movies);

  useEffect(() => {
    loadMovies();
  }, [loadMovies]);

  useEffect(() => {
    setFilteredMovies(movies);
  }, [movies]);

  const paginatedMovies = useMemo(() => {
    const startIndex = (currentPage - 1) * itemsPerPage;
    const endIndex = startIndex + itemsPerPage;
    return filteredMovies.slice(startIndex, endIndex);
  }, [filteredMovies, currentPage, itemsPerPage]);

  useEffect(() => {
    changePage(1);
  }, [filteredMovies, changePage]);

  if (status === MovieStatusEnum.FAILED) {
    return <p>Error loading movies: {error}</p>;
  }

  return (
    <div
      className={`flex flex-col h-screen p-7 bg-slate-50 text-sm ${status === MovieStatusEnum.LOADING && 'justify-center'}`}
    >
      <div className="flex flex-col overflow-hidden">
        {status === MovieStatusEnum.LOADING ? (
          <Loader />
        ) : (
          <>
            <h1 className="text-center font-bold text-2xl p-3">Movie Management System</h1>
            <CategoryFilter onChange={setFilteredMovies} />
            <MoviesList movies={paginatedMovies} />
            <Pagination total={filteredMovies.length} />
          </>
        )}
      </div>
    </div>
  );
};

export default App;
