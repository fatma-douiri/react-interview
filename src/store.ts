import { configureStore } from '@reduxjs/toolkit';

import moviesSlice from './features/movies/moviesSlice';

export const store = configureStore({
  devTools: true,
  reducer: {
    movies: moviesSlice,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
