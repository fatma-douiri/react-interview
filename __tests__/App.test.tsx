import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { describe, expect, it, vi } from 'vitest';

import App from '../src/App';
import { store } from '../src/store';

vi.useFakeTimers();

describe('Renders main page correctly', async () => {
  it('Should render the page correctly', async () => {
    const container = render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
