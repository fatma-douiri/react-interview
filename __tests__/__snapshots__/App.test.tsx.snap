// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders main page correctly > Should render the page correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col h-screen p-7 bg-slate-50 text-sm justify-center"
      >
        <div
          class="flex flex-col overflow-hidden"
        >
          <div
            class="flex flex-grow items-center justify-center"
          >
            <svg
              aria-hidden="true"
              aria-label="spinner"
              class="custom-spinner max-md:w-16 max-md:h-16 w-40 h-40 text-gray-200 animate-spin dark:text-gray-600 fill-gray-600"
              fill="currentColor"
              height="24"
              viewBox="0 0 100 101"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M100 50.59c0 27.615-22.386 50.001-50 50.001s-50-22.386-50-50 22.386-50 50-50 50 22.386 50 50m-90.919 0c0 22.6 18.32 40.92 40.919 40.92s40.919-18.32 40.919-40.92c0-22.598-18.32-40.918-40.919-40.918S9.081 27.992 9.081 50.591"
                fill="currentColor"
              />
              <path
                d="M93.968 39.04c2.425-.636 3.894-3.128 3.04-5.486A50 50 0 0 0 41.735 1.279c-2.474.414-3.922 2.919-3.285 5.344s3.12 3.849 5.6 3.484a40.916 40.916 0 0 1 44.131 25.769c.902 2.34 3.361 3.802 5.787 3.165"
                fill="currentFill"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col h-screen p-7 bg-slate-50 text-sm justify-center"
    >
      <div
        class="flex flex-col overflow-hidden"
      >
        <div
          class="flex flex-grow items-center justify-center"
        >
          <svg
            aria-hidden="true"
            aria-label="spinner"
            class="custom-spinner max-md:w-16 max-md:h-16 w-40 h-40 text-gray-200 animate-spin dark:text-gray-600 fill-gray-600"
            fill="currentColor"
            height="24"
            viewBox="0 0 100 101"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.59c0 27.615-22.386 50.001-50 50.001s-50-22.386-50-50 22.386-50 50-50 50 22.386 50 50m-90.919 0c0 22.6 18.32 40.92 40.919 40.92s40.919-18.32 40.919-40.92c0-22.598-18.32-40.918-40.919-40.918S9.081 27.992 9.081 50.591"
              fill="currentColor"
            />
            <path
              d="M93.968 39.04c2.425-.636 3.894-3.128 3.04-5.486A50 50 0 0 0 41.735 1.279c-2.474.414-3.922 2.919-3.285 5.344s3.12 3.849 5.6 3.484a40.916 40.916 0 0 1 44.131 25.769c.902 2.34 3.361 3.802 5.787 3.165"
              fill="currentFill"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
